toolchain/lower/context.cpp

```diff
@@ -0,0 +1,88 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/context.h"
+
+#include "common/check.h"
+#include "common/vlog.h"
+#include "llvm/Transforms/Utils/ModuleUtils.h"
+#include "toolchain/lower/file_context.h"
+#include "toolchain/sem_ir/inst_namer.h"
+
+namespace Carbon::Lower {
+
+Context::Context(llvm::LLVMContext& llvm_context,
+                 llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs,
+                 std::optional<llvm::ArrayRef<Parse::GetTreeAndSubtreesFn>>
+                     tree_and_subtrees_getters_for_debug_info,
+                 llvm::StringRef module_name, llvm::raw_ostream* vlog_stream)
+    : llvm_context_(&llvm_context),
+      llvm_module_(std::make_unique<llvm::Module>(module_name, llvm_context)),
+      file_system_(std::move(fs)),
+      di_builder_(*llvm_module_),
+      di_compile_unit_(
+          tree_and_subtrees_getters_for_debug_info
+              ? BuildDICompileUnit(module_name, *llvm_module_, di_builder_)
+              : nullptr),
+      tree_and_subtrees_getters_for_debug_info_(
+          tree_and_subtrees_getters_for_debug_info),
+      vlog_stream_(vlog_stream) {}
+
+auto Context::GetFileContext(const SemIR::File* file,
+                             const SemIR::InstNamer* inst_namer)
+    -> FileContext* {
+  auto insert_result = file_contexts_.Insert(file->check_ir_id(), [&] {
+    auto file_context =
+        std::make_unique<FileContext>(*this, *file, inst_namer, vlog_stream_);
+    file_context->PrepareToLower();
+    return file_context;
+  });
+  return insert_result.value().get();
+}
+
+auto Context::Finalize() -> std::unique_ptr<llvm::Module> {
```

Comment:
```suggestion
auto Context::Finalize() && -> std::unique_ptr<llvm::Module> {
```
This method seems destructive to `this`, is the class meant to be used after calling Finalize?